Add live sync of site styles (block theme styles) to the email editor.

triple0t opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes #58964 WOOPLUG-4691

(For Bug Fixes) Bug introduced in PR # .

This PR implements live synchronization of site styles to email templates, allowing email designs to automatically inherit and adapt the site's theme styles without manual configuration in the email editor.

**Why**

Email templates in the Email Editor currently utilize default styles that don't match the site's appearance, resulting in visual inconsistency between the website and email communications. Users have to manually configure email styles to match their site, which is time-consuming and often incomplete.

The goal is to automatically sync site theme styles (from theme.json + user customizations via site editor) to emails, ensuring visual consistency across all customer touchpoints without requiring manual setup.

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| Before | After |
| ------ | ----- |
|     <img width="877" height="963" alt="Screenshot 2025-07-18 at 1 24 20 AM" src="https://github.com/user-attachments/assets/7346b8fe-5fdc-4acd-9279-b14abc9e5c8d" /> |     <img width="1081" height="985" alt="Screenshot 2025-07-18 at 1 23 45 AM" src="https://github.com/user-attachments/assets/b964b186-d6e4-45f1-9941-a55edfd686de" />  |
|  <img width="1009" height="988" alt="Screenshot 2025-07-18 at 1 24 30 AM" src="https://github.com/user-attachments/assets/10584940-5cdd-4baf-b854-3639e1800b58" /> | <img width="992" height="885" alt="Screenshot 2025-07-18 at 1 23 39 AM" src="https://github.com/user-attachments/assets/bdeb72e2-7bd6-407b-a68a-aaa5a0889df6" /> | 


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:


1. **Basic Sync Test**:
   - Switch to a block theme with custom colors/typography
   - Create/edit an email template
   - Verify email preview reflects site theme styles

2. **User Customization Test**:
   - Go to Appearance > Editor > Styles
   - Modify colors, typography, or spacing
   - Check that email templates automatically update

3. **Override Test**:
   - Set custom styles directly in email editor
   - Verify these override the synced site styles
   - Confirm site style changes don't override manual email settings

4. **Filter Test**:
   ```php
   // Disable sync
   add_filter( 'woocommerce_email_editor_site_style_sync_enabled', '__return_false' );
   
   // Custom style modifications
   add_filter( 'woocommerce_email_editor_synced_site_styles', function( $synced_data, $site_data ) {
       $synced_data['styles']['color']['background'] = '#ffffff';
       return $synced_data;
   }, 10, 2 );
   ```


<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

triple0t committed:
Move button styling from Core/Theme.json to Engine/Theme.json

The Core/Theme.json has higher priority than the Engine/Theme.json. This affects styles set by the user and prevents them from working.

triple0t committed:
Fix global styles preview panel throwing errors for classic themes

triple0t committed:
Add Site Style Sync Controller to Email Editor

Integrate the Site Style Sync Controller into the Email Editor Container and Theme Controller. This allows for the merging of synced styles from the current active theme, enhancing the theme management capabilities.

triple0t committed:
Add Site Style Sync Controller for Email Templates

Introduce the Site Style Sync Controller to manage the synchronization of site styles with email templates. This implementation ensures visual consistency by converting site theme styles into email-compatible formats, enhancing the email editor's functionality.

triple0t committed:
Add utility methods for CSS value conversion in Styles Helper

Enhance the Styles Helper class by adding methods to convert CSS values to static pixel values, remove CSS units, and handle clamp() values. This improves compatibility for email clients and provides better styling options for email templates. Additionally, update unit tests to validate the new functionality.

triple0t committed:
Remove unnecessary items from the Site Style Sync Controller  and reuse the Styles_Helper

triple0t committed:
Add tests for the Site Style Sync Controller  class and update Theme_Controller_Test to account for site theme sync

triple0t committed:
Update documentation

triple0t committed:
Enhance color palette handling in Preview component to ensure default colors are safely concatenated with theme colors, preventing potential errors when default colors are undefined.

triple0t committed:
Add changelog files

triple0t committed:
Apply PR review suggestion

costasovo commented on packages/php/email-editor/src/Integrations/Utils/class-styles-helper.php:
```diff
@@ -43,4 +43,88 @@ public static function parse_styles_to_array( string $styles ): array {
+	 * @param ?string $fallback The fallback value to use if the input is not a valid CSS value.
+	 * @return ?string The static pixel value (e.g., 30px).
+	 */
+	public static function convert_to_px( string $input, $fallback = '16px' ): ?string {
```
This is really nice addition!
The description and method name suggest it converts any value, but seeing that percentage uses 16px as a base, I think it would be nice to make it more obvious that the method is intended only for font size and spacing.
Or perhaps could we use the $fallback (I guess renamed to $base) as a base value?

triple0t replied:
Implemented in https://github.com/woocommerce/woocommerce/pull/59757/commits/f5071603be4846219bf28bd26d3872b08a1e211d

costasovo commented on packages/php/email-editor/src/Engine/class-site-style-sync-controller.php:
```diff
@@ -225,17 +209,17 @@ private function convert_color_styles( array $color_styles ): array {
+		// Preserve font family.
 		if ( isset( $typography_styles['fontFamily'] ) ) {
-			$email_typography['fontFamily'] = $this->convert_to_email_safe_font( $typography_styles['fontFamily'] );
+			$email_typography['fontFamily'] = $typography_styles['fontFamily'];
```
Just curious. Why did you decide to drop the conversion to the safe font family? I thought it was a good idea.

triple0t replied:
This is because managing the font family would be a headache. We would need to attempt mapping different types of font families to an email-safe font. Plus, the font map might already exist in the editor, and we need to check for duplicates, remove them, etc.
In this case, I decided to proceed with the current default

costasovo replied:
@triple0t I see, and it makes sense not to include font mapping in this PR, but I see there is a significant WYSIWYG difference when a web uses a special font. 

### Email Editor
<img width="673" height="1086" alt="Screenshot 2025-07-22 at 10 17 40" src="https://github.com/user-attachments/assets/f13437c3-4890-484e-9ce7-9fac8eaa1320" />

### Rendered output
<img width="715" height="1323" alt="Screenshot 2025-07-22 at 10 17 49" src="https://github.com/user-attachments/assets/fd8bcca1-16ee-4e5e-9af6-ea79bb91b669" />

So I think we should revisit this and somehow attempt to map site's font to some of our supported fonts. In my opinion the WYSIWYG experience is more important than the chance that we don't pick the most similar font. What do you think?

triple0t replied:
Implemented in https://github.com/woocommerce/woocommerce/pull/59757/commits/872b5aeeb3bcee7eb9ee4646fb0d8cc89b3a9760

costasovo reviewed (changes requested):
Great job! 
I really appreciate the documentation and the test coverage.
Also, kudos for improving the style helpers.

I added a couple of comments.

When testing I noticed that button styles are out of sync between editor, site and rendered output. I didn't dig to find the cause but I see some changes about buttons in the PR.

**Button in site editor**
<img width="382" height="539" alt="Screenshot 2025-07-18 at 12 44 45" src="https://github.com/user-attachments/assets/b7bc4072-e6dc-47dd-ba56-33794cab958d" />
**Button in email editor - no background applied**
I expected it to default to the same style as site.
<img width="195" height="286" alt="Screenshot 2025-07-18 at 12 45 13" src="https://github.com/user-attachments/assets/4a4097ce-87b8-4a27-a961-090e7d4ab4b8" />
**Button in renderer output**
This is probably the default button style applied from email theme
<img width="308" height="313" alt="Screenshot 2025-07-18 at 12 45 03" src="https://github.com/user-attachments/assets/a376cd8e-5b32-477d-a6b4-4aa9df2c6c90" />

Note: If you prefer, we can deal with the buttons in a separate issue.

triple0t committed:
Update convert_to_px method to use a different base font size when provided

triple0t commented:
Thank you for the review, @costasovo 

> When testing I noticed that button styles are out of sync between editor, site and rendered output. I didn't dig to find the cause but I see some changes about buttons in the PR.
> Note: If you prefer, we can deal with the buttons in a separate issue.

I noticed this as well, and I think it's due to having an empty button object in your global styles.
Please check the DB post table for `wp-global-styles-woocommerce-email` and confirm the post content contains an empty button object.
You can delete the entry and refresh the page.

costasovo reviewed (approved):
@triple0t Looks good. Great job!

> I noticed this as well, and I think it's due to having an empty button object in your global styles.
Please check the DB post table for wp-global-styles-woocommerce-email and confirm the post content contains an empty button object.
You can delete the entry and refresh the page.

In this second round of testing I was starting with a fresh env and the issue was not present. So it was probably related to incorrect global styles post.

> I'm not sure if I should attempt to fix that particular issue in this PR or not (given it's already complex scope). Please let me know your thoughts.

I think we don't need to address it in this PR. Do we actually need to address it? I mean if a user makes an edit in email styles wouldn't that fix the issue as well? 

#### Testing
I tried to change styles in the site editor and observed that the change was reflected in the email.
I see there are different fonts in rendered output vs editor but this is out of scope for this ticket.
**Settings in the style editor**
<img width="1236" height="1280" alt="Screenshot 2025-07-22 at 10 17 59" src="https://github.com/user-attachments/assets/b2e4a240-9217-4501-ad2a-67ea382bafa9" />
**Email in the editor**
<img width="673" height="1086" alt="Screenshot 2025-07-22 at 10 17 40" src="https://github.com/user-attachments/assets/79550b1d-469f-433c-b1a4-a3b864085b1a" />
**Rendered Email**
<img width="715" height="1323" alt="Screenshot 2025-07-22 at 10 17 49" src="https://github.com/user-attachments/assets/c16d8c81-b83c-45da-84c9-f820f698a4ea" />

triple0t committed:
Add email-safe font handling to Site Style Sync Controller

Why: to ensure a WYSIWYG experience between the email editor and the rendered output.

Introduce functionality to retrieve and convert font families to email-safe alternatives. This includes a new method to fetch email-safe fonts from theme.json and updates to typography style conversion to ensure compatibility with email clients. Additionally, comprehensive tests have been added to validate the new font handling features.
